docs/en/integrations/data-ingestion/data-formats/templates-regex.md

```diff
@@ -0,0 +1,237 @@
+```
+
+We can use a [Template](/docs/en/interfaces/formats.md/#format-template) format to import this data. We have to define a template string with values placeholders for each row of input data:
+
+```
+<time> [error] client: <ip>, server: <host> "<request>"
+```
+
+Let's create a table to import our data into:
+```sql
+CREATE TABLE error_log
+(
+    `time` DateTime,
+    `ip` String,
+    `host` String,
+    `request` String
+)
+ENGINE = MergeTree
+ORDER BY (host, request, time)
+```
+
+To import data using a given template, we have to save our template string in a file ([row.template](assets/row.template) in our case):
+```
+${time:Escaped} [error]  client: ${ip:CSV}, server: ${host:CSV} ${request:JSON}
+```
+
+We define a name of a column and escaping rule in a `${name:escaping}` format. Multiple options are available here, like CSV, JSON, Escaped, or Quoted, which implement [respective escaping rules](/docs/en/interfaces/formats.md/#format-template).
+
+Now we can use the given file as an argument to the `format_template_row` settings option while importing data (*note, that template and data files **should not have** an extra `\n` symbol at the end of file*):
+
+```sql
+INSERT INTO error_log FROM INFILE 'error.log'
+SETTINGS format_template_row = 'row.template'
+FORMAT Template
+```
+
+And we can make sure our data was loaded into the table:
+
+```sql
+SELECT
+    request,
+    count(*)
+FROM error_log
+GROUP BY request;
+
+┌─request──────────────────────────────────────────┬─count()─┐
+│ GET /img/close.png HTTP/1.1                      │     176 │
+│ GET /h5/static/cert/icon_yanzhengma.png HTTP/1.1 │     172 │
+│ GET /phone/images/icon_01.png HTTP/1.1           │     139 │
+│ GET /apple-touch-icon-precomposed.png HTTP/1.1   │     161 │
+│ GET /apple-touch-icon.png HTTP/1.1               │     162 │
+│ GET /apple-touch-icon-120x120.png HTTP/1.1       │     190 │
+└──────────────────────────────────────────────────┴─────────┘
+```
+
+### Skipping whitespaces
+Consider using [TemplateIgnoreSpaces](/docs/en/interfaces/formats.md/#templateignorespaces), which allows skipping whitespaces between delimiters in a template:
+```
+Template:               -->  "p1: ${p1:CSV}, p2: ${p2:CSV}"
+TemplateIgnoreSpaces    -->  "p1:${p1:CSV}, p2:${p2:CSV}"
+```
+
+## Exporting data using templates
+
+We can also export data to any text format using templates as well. In this case, we have to create two files:
+
+[Result set template](assets/output.results), which defines the layout for the whole result set:
+```
+== Top 10 IPs ==
+${data}
+--- ${rows_read:XML} rows read in ${time:XML} ---
+```
+
+Here, `rows_read` and `time` are system metrics available for each request. While `data` stands for generated rows (`${data}` should always come as a first placeholder in this file), based on a template defined in a [**row template file**](assets/output.rows):
```

Comment:
This suggests `${data}` has to always be the first use of a variable? is this a strict requirement

Reply:
yep, documented and confirmed, ${data} should come first or error